dali/kernels/reduce/mean_stddev_gpu_impl.cuh

```diff
@@ -0,0 +1,386 @@
+  using ReduceBase = ReduceImplGPU<Out, In, Acc, StdDevImplGPU<Out, In, Mean, Acc>>;
+  using VarBase = VarianceImplBase<Out, In, Mean, StdDevImplGPU<Out, In, Mean, Acc>>;
+  using MeanBase = RootMeanImplBase<Out, In, StdDevImplGPU<Out, In, Mean, Acc>>;
+
+  using Preprocessor = typename VarBase::Preprocessor;
+  template <int non_reduced_dims>
+  using PreprocessorBank = typename VarBase::template PreprocessorBank<non_reduced_dims>;
+
+  using Postprocessor = typename MeanBase::Postprocessor;
+  reductions::sum GetReduction() const { return {}; }
+
+  void Run(KernelContext &kctx,
+           const OutListGPU<Out> &out,
+           const InListGPU<In> &in,
+           const InListGPU<Mean> &mean) {
+    this->InitMean(mean);
+    ReduceBase::Run(kctx, out, in);
+  }
+};
+
+template <typename Out, typename ScaleAndReg>
+struct RegularizedInvSqrt {
+  ScaleAndReg scale = 1, reg = 0;
+
+  template <typename T>
+  DALI_HOST_DEV Out operator()(T x) const {
+    float s = scale * x + reg;
+    return s ? ConvertSat<Out>(rsqrt(s)) : Out(0);
+  }
+};
+
+template <typename Out, typename In, typename Actual>
+class RegularizedInvRMS {
+ public:
+  Actual &This() { return static_cast<Actual&>(*this); }
+  const Actual &This() const { return static_cast<const Actual&>(*this); }
+
+  using param_t = std::conditional_t<std::is_same<Out, double>::value, double, float>;
+  using Postprocessor = RegularizedInvSqrt<Out, param_t>;
+
+  void SetRegularizationTerm(param_t reg) {
+    regularization_ = reg * reg;
+  }
+
+  param_t regularization_ = 0.0f;
+
+  Postprocessor *GetPostprocessorsImpl(WorkArea &wa) const {
+    assert(!This().ReduceBatch());
+    int n = This().SimplifiedOutputShape().num_samples();
+    Postprocessor *pp = wa.ParamBuffer<Postprocessor>(n);
+    for (int i = 0; i < n; i++) {
+      DALI_ENFORCE(This().ReducedElements(i) > 0, "Cannot calculate a mean from 0 elements");
+      pp[i].scale = param_t(1.0 / This().ReducedElements(i));
+      pp[i].reg = regularization_;
+    }
+    return pp;
+  }
+
+  Postprocessor GetPostprocessorImpl() const {
+    assert(This().ReduceBatch());
+    DALI_ENFORCE(This().TotalReducedElements() > 0, "Cannot calculate a mean from 0 elements");
+    return { param_t(1.0f / This().TotalReducedElements()), regularization_ };
+  }
+};
+
+template <typename Out, typename In, typename Mean = Out, typename Acc = Out>
+class InvStdDevImplGPU :
+      public ReduceImplGPU<Out, In, Acc, InvStdDevImplGPU<Out, In, Mean, Acc>>,
+      public VarianceImplBase<Out, In, Mean, InvStdDevImplGPU<Out, In, Mean, Acc>>,
+      public RegularizedInvRMS<Out, In, InvStdDevImplGPU<Out, In, Mean, Acc>> {
+ public:
+  using ReduceBase = ReduceImplGPU<Out, In, Acc, InvStdDevImplGPU<Out, In, Mean, Acc>>;
+  using VarBase = VarianceImplBase<Out, In, Mean, InvStdDevImplGPU<Out, In, Mean, Acc>>;
+  using MeanBase = RegularizedInvRMS<Out, In, InvStdDevImplGPU<Out, In, Mean, Acc>>;
+
+  using Preprocessor = typename VarBase::Preprocessor;
```

Comment:
When do you need to just use `using Preprocessor = ...` and when do you provide the GetPrerocessorImpl & co? 
Is it beacause one of the ImplBase/Base classes (btw, RegularizedInvRMS doesn't have Base in it's name) already provide them, and you need to bring only the types to scope here?

Maybe going a bit into the design of those classes, I like the customization points for Pre & Post processors, but I wonder, if it wouldn't be easier to read if they were just explicitly defined.

Basically those ImplBase classes represent a non-documented concept. It would be really helpful if there was a doc explaining, that you can provide such class, that needs to provide:
P GetPreprocessor(...) // for batch reduction
P* GetPreprocessor() // for per sample reduction
...
Where `P` is expected to provide call operator.
Such "Concept"/Policy class could be passed to ReduceImplGpu, and it would access it's Pre/Post processors directly without a chain of unholy overloads.

Reply:
The problem is, that the base class ReduceImplBase also defines Preprocessor, Postprocessor, etc - we need to shadow them properly here; without this using, there's ambiguity, because the type is defined in multiple base classes.